src/install/dracut-install.c

```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
+                                                return -1;
+
+                                        strncat(supplier_path, supplier_slice, sizeof(supplier_path) - strlen(supplier_path) - 1);
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, strlen(supplier_path), &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+
+                                                        int err = hashmap_put(suppliers, supplier_module, supplier_module);
+                                                        if (err < 0) {
+                                                                free(supplier_module);
+                                                                return err;
+                                                        }
+                                                }
+                                        }
+                                }
+                        }
+                }
+                closedir(d);
+        }
+
+        return 0;
+}
+
+static int find_suppliers_for_module(struct kmod_ctx *ctx, Hashmap *suppliers, const char *module)
+{
+        int err = 0;
+
+        _cleanup_fts_close_ FTS *fts;
+        char *paths[] = { "/sys/devices/platform", NULL };
+        fts = fts_open(paths, FTS_NOSTAT | FTS_PHYSICAL, NULL);
+
+        for (FTSENT *ftsent = fts_read(fts); ftsent != NULL; ftsent = fts_read(fts)) {
+                if (strcmp(ftsent->fts_name, "modalias") == 0) {
+                        FTSENT *sysfs_node = ftsent->fts_parent;
+
+                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                        struct kmod_list *l = NULL;
+
+                        if (find_module_from_sysfs_node(ctx, sysfs_node->fts_path, sysfs_node->fts_pathlen, &list) < 0)
+                                continue;
+
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        err = find_suppliers_for_sys_node(ctx, suppliers, sysfs_node->fts_path, sysfs_node->fts_pathlen);
+                                        if (err)
+                                                return err;
```

Comment:
am I wrong to think this *always* overlaps with the first iteration of the loop below?